File: src/app/class/fight/fight.ts
import { Monster } from '../monster/monster';

export class Fight {
    id: number;
    name: string;
    members: Monster[];
    challengeRating: number;

    constructor(props) {
        this.id = props.id;
        this.name = props.name;
        this.challengeRating = props.challengeRating;

        if (props.members) {
            this.members = props.members.map(member => new Monster(member));
        }
    }
}

File: campaign_manager/gos/serializers/fight.py
from rest_framework import serializers

from gos.models.fight import Fight
from gos.models.monster import Monster
from gos.serializers.fight_member import FightMemberSerializer


class FightSerializer(serializers.ModelSerializer):
    members = FightMemberSerializer(many=True)

    class Meta:
        model = Fight
        fields = [
            'id',
            'name',
            'members',
            'challenge_rating',
        ]
    
    def create(self, validated_data):
        members_data = validated_data.pop('members')
        fight = Fight.objects.create(**validated_data)
        for member in members_data:
            monster = Monster.objects.get(name=member['monster']['name'])
            fight.members.add(
                monster,
                through_defaults={'initiative_order': member['initiative_order']}
            )
        return fight
